### PR TITLE
Create GC scan type for flattened arrays

### PR DESCRIPTION
**Pull request by youngar**
Add a new scan type for flattened arrays. I want to make sure that everyone agrees with this approach. There are many places which will need to be updated to support the new scan type, but for now I have only changed MarkingDelegate and ScavengerDelegate.  The places which we will need to update are:
```
DDR_VM/src/com/ibm/j9ddr/vm29/j9/gc/GCObjectModel_V1.java
gc_api/HeapIteratorAPI.cpp
gc_base/ReferenceChainWalker.cpp
gc_glue_java/CompactDelegate.cpp
gc_glue_java/CompactSchemeFixupObject.cpp
gc_glue_java/MetronomeDelegate.hpp
gc_glue_java/ObjectIterator.hpp
gc_glue_java/MarkingDelegate.hpp
gc_glue_java/ObjectModel.hpp
gc_glue_java/ScavengerDelegate.cpp
gc_realtime/RealtimeAccessBarrier.cpp
gc_trace_vlhgc/TgcInterRegionReferences.cpp
gc_vlhgc/GlobalMarkCardScrubber.cpp
gc_vlhgc/GlobalMarkCardScrubber.hpp
gc_vlhgc/IncrementalGenerationalGC.cpp
gc_vlhgc/WriteOnceCompactor.cpp
gc_vlhgc/CopyForwardScheme.cpp
gc_vlhgc/GlobalMarkingScheme.cpp
gc_vlhgc/PartialMarkingScheme.cpp
```

cc @dmitripivkine @amicic

### CODE REVIEW

File: runtime/gc_glue_java/MarkingDelegate.hpp [hunk 1]
```diff
@@ -134,6 +134,12 @@ class MM_MarkingDelegate
 			objectScanner = GC_PointerArrayObjectScanner::newInstance(env, objectPtr, scannerSpace, GC_ObjectScanner::indexableObject, slotsToDo, startIndex);
 			break;
 		}
+		case GC_ObjectModel::SCAN_FLATTENED_ARRAY_OBJECT:
+		{
+			Assert_MM_true(J9_IS_J9CLASS_FLATTENED(clazzPtr));
+			Assert_MM_unreachable();
```
Comment (dmitripivkine) [inline runtime/gc_glue_java/MarkingDelegate.hpp]: technically `Assert_MM_unimplemented()` fits better

**Comment by dmitripivkine** [comment]
> Add a new scan type for flattened arrays. I want to make sure that everyone agrees with this approach. There are many places which will need to be updated to support the new scan type, but for now I have only changed MarkingDelegate and ScavengerDelegate.

Sounds correct unfortunately. Everywhere object type switch exists it should be added. I am not sure there is simple way to reduce number of places to be modified.

We can do changes in order to support GC policies one by one. For example if new scanner would be available we can start with `optthruput`. However if it is problematic we can start with `balanced` (where old style iterators are used) and restrict it even more by disabling PGC for first implementation

Some of places used in debug mode only (for example `MM_CompactDelegate::verifyHeap()`) and might be implemented with low priority.

Also for first cut of implementation we can disable Array Split for Flattened types (both for Mark and Scavenge)

**Comment by youngar** [comment]
I think I will have to make sure that no code is asserting with the new scan type before we can accept this.  The added assertions will probably cause the existing valuetype tests to fail.

**Comment by youngar** [comment]
Looks like the tests were triggering the assertion in `MarkingDelegate`with `gencon`.  When turning off GC support `-Xgcpolicy:nogc`, we hit the following assertion:

```
MM_StandardAccessBarrier::preObjectRead (this=0x7ffff00477f0, vmThread=0x14200, srcObject=<optimized out>, srcAddress=0x60eeb7f6c) at StandardAccessBarrier.cpp:845
845		Assert_GC_true_with_message(env, !_extensions->scavenger->isObjectInEvacuateMemory((omrobjectptr_t)srcAddress) || _extensions->isScavengerBackOutFlagRaised(), "readObject %llx in Evacuate\n", srcAddress);
```
I'm guessing we shouldn't be going through the standard access barrier when were in `nogc` mode
```
(gdb) p this->_extensions->scavenger
$2 = (MM_Scavenger *) 0x0
```

**Comment by amicic** [comment]
I'm guessing that read barrier type is properly check, but there are paths that would invoke the preObjectBarrier that would call it without check the barrier type. Macros like J9OBJECT_OBJECT_LOAD do check, but possibly slow path, but there are other ways to trigger the barrier.

Either those paths should be fixed to check the type or the barrier itself check against scavenge pointer. 

What's stacktrace?

**Comment by youngar** [comment]
```
#0  MM_StandardAccessBarrier::preObjectRead (this=0x7ffff00477f0, vmThread=0x14200, srcObject=<optimized out>, srcAddress=0x60eeb7f6c) at StandardAccessBarrier.cpp:845
#1  0x00007ffff609fd85 in MM_ObjectAccessBarrier::mixedObjectReadObject (this=0x7ffff00477f0, vmThread=0x14200, srcObject=0x60eeb7f50, srcOffset=<optimized out>, isVolatile=<optimized out>) at ObjectAccessBarrier.cpp:573
#2  0x00007ffff609f96d in MM_ObjectAccessBarrier::copyObjectFields (this=0x7ffff00477f0, vmThread=0x14200, objectClass=0x370700, srcObject=0x60eeb7f50, srcOffset=8, destObject=0x60eeb8318, destOffset=4) at ObjectAccessBarrier.cpp:1436
#3  0x00007ffff609e622 in MM_ObjectAccessBarrier::copyObjectFieldsFromArrayElement (this=0x7ffff00477f0, vmThread=0x14200, arrayClazz=<optimized out>, destObject=0x60eeb8318, arrayRef=0x60eeb7f50, index=<optimized out>) at ObjectAccessBarrier.cpp:1143
#4  0x00007ffff72439ca in MM_ObjectAccessBarrierAPI::copyObjectFieldsFromArrayElement (index=<optimized out>, arrayRef=<optimized out>, destObject=<optimized out>, arrayClazz=<optimized out>, vmThread=<optimized out>, this=<optimized out>) at ../gc_include/ObjectAccessBarrierAPI.hpp:392
#5  VM_BytecodeInterpreter::aaload (_pc=<optimized out>, _sp=<optimized out>, this=<optimized out>) at BytecodeInterpreter.hpp:5843
#6  VM_BytecodeInterpreter::run (this=0x7ffff00477f0, this@entry=0x7ffff7bee680, vmThread=0x7ffff003b230) at BytecodeInterpreter.hpp:9892
#7  0x00007ffff7238c20 in bytecodeLoop (currentThread=<optimized out>) at BytecodeInterpreter.cpp:109
#8  0x00007ffff72f74e2 in c_cInterpreter () at xcinterp.s:160
#9  0x00007ffff72a8d0e in sidecarInvokeReflectMethodImpl (currentThread=currentThread@entry=0x14200, methodRef=<optimized out>, recevierRef=<optimized out>, argsRef=<optimized out>) at callin.cpp:1149
#10 0x00007ffff72a96f5 in sidecarInvokeReflectMethod (currentThread=0x14200, methodRef=<optimized out>, recevierRef=<optimized out>, argsRef=<optimized out>) at callin.cpp:1293
#11 0x00007ffff5ec4874 in JVM_InvokeMethod_Impl (env=0x14200, method=0x109350, obj=0x109348, args=0x109340) at sunvmi.c:354
```

**Comment by youngar** [comment]
Its crashing in code related to value types, on this line BytecodeInterpreter.hpp:5843 `_objectAccessBarrier.copyObjectFieldsFromArrayElement`
```c++
#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
				J9Class *arrayrefClass = J9OBJECT_CLAZZ(_currentThread, arrayref);
				if (J9_IS_J9CLASS_FLATTENED(arrayrefClass)) {
					j9object_t newObjectRef = _objectAllocate.inlineAllocateObject(_currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, false, false);

					if (NULL == newObjectRef) {
						buildGenericSpecialStackFrame(REGISTER_ARGS, 0);
						pushObjectInSpecialFrame(REGISTER_ARGS, arrayref);
						updateVMStruct(REGISTER_ARGS);
						newObjectRef = _vm->memoryManagerFunctions->J9AllocateObject(_currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, J9_GC_ALLOCATE_OBJECT_NON_INSTRUMENTABLE);
						VMStructHasBeenUpdated(REGISTER_ARGS);
						arrayref = popObjectInSpecialFrame(REGISTER_ARGS);
						restoreGenericSpecialStackFrame(REGISTER_ARGS);
						if (J9_UNEXPECTED(NULL == newObjectRef)) {
							rc = THROW_HEAP_OOM;
							return rc;
						}
						arrayrefClass = VM_VMHelpers::currentClass(arrayrefClass);
					}

					_objectAccessBarrier.copyObjectFieldsFromArrayElement(_currentThread, arrayrefClass, newObjectRef, (J9IndexableObject *) arrayref, index);
					value = newObjectRef;
				} else
#endif /* if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
```

**Comment by amicic** [comment]
I guess, the easiest is to check if scavenge pointer is not null inside preObjectRead. This is what is already done in a few other variants of that call (same method name, just different signature).

**Comment by youngar** [comment]
Created a pull request for this here: https://github.com/eclipse/openj9/pull/7939

**Comment by youngar** [comment]
@dmitripivkine @amicic With the prerequisite commit #7939 merged, this change is good to go now. I don't think we need to strictly wait for https://github.com/eclipse/openj9/pull/7949 since the test is not run in automated testing.

**Comment by dmitripivkine** [comment]
There are no flattened arrays in regular build. I believe travis-ci testing is sufficient enough and it can be merged as soon as all approvals received

**Comment by youngar** [comment]
Thanks @dmitripivkine

**Comment by knn-k** [comment]
I opened Issue #7960 caused by this PR.  Please fix.  Thanks.

**Comment by pshipton** [comment]
The fix is in #7956

**Pull request merged**